src/hotspot/share/runtime/interfaceSupport.inline.hpp

```diff
@@ -202,7 +202,9 @@ class ThreadInVMfromNative : public ThreadStateTransition {
     trans_from_native(_thread_in_vm);
   }
   ~ThreadInVMfromNative() {
-    trans(_thread_in_vm, _thread_in_native);
+    assert(_thread->thread_state() == _thread_in_vm, "coming from wrong thread state");
+    assert(!_thread->has_last_Java_frame() || _thread->frame_anchor()->walkable(), "Unwalkable stack in vm->native transition");
+    _thread->set_thread_state(_thread_in_native);
```

Comment:
I looked over the transitions this fall, I don't know but there are some of these oddities.
One problem is the some of the generic named transitions RAII are actually supposed to be used only at certain places and some optimizations have been added.
E.g. in this case someone have figured out that we can remove make_walkable(), but as a generic rule when going to a safe state (blocked/native)  you should call make_walkable. My guess is that calling make_walkable on every JNI/JVM TI call was expensive on Solaris, so someone optimized that.
Now I'm unsure if that is actually true (I my feeling is no, because we could accidentally call Java if we hit e.g. class loading or something). And the optimization is no longer valid, make_walkable() is light weight on current hardware AFAIK.

I updated to what I see as a correct transition to an safe state without any short cuts.
I also notice I missed an assert from the previous code.

Re-running all test, since these kinds of changes should not be taken lightly.